### PR TITLE
Log User Actions (#392)

### PR DESCRIPTION
**Pull request by phase**
Logs have been added for (in order they appeared on GitHub):

* Editing a project page
* Uploading a new project
* Flagging a project
* Uploading a new icon to a project
* Removing an icon from a project
* Removing a user's membership from a project
* Editing project settings
* Renaming a project
* Changing the visibility of a project
* Publishing a project
* Setting a project as `NeedsApproval`
* Deleting a project
  * Soft-deleting a project
* Editing the description of a version
* Setting the Recommended Version of a project to a version
* Approving a version
* Uploading a version
* Publishing a version
* Deleting a version

Actions are stored in a `user_actions_log` table with the schema:

```sql
  CREATE TABLE user_action_log (
    id         SERIAL PRIMARY KEY,
    created_at TIMESTAMP NOT NULL DEFAULT NOW(),
    user_id    INT REFERENCES users,
    address    INET NOT NULL,
    action     TEXT
  );
```

"Logging" any of these actions with print them out to STDOUT with the `info` level, and add them to the table. 

(#392)

### CODE REVIEW

File: app/models/user/UserAction.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import db.impl.UserActionLogTable
+import db.impl.model.OreModel
+import ore.user.UserOwned
+
+case class UserAction(override val id: Option[Int] = None,
+                      override val createdAt: Option[Timestamp] = None,
+                      private val _userId: Int,
+                      private val _action: String
+                     )
```
Comment (phase) [inline app/models/user/UserAction.scala]: this also needs to include the IP

**Comment by phase** [comment]
This is good to go. If you have anything else that you want to be logged just ask @mbax

**Comment by felixoi** [comment]
Needs UI.

[https://gyazo.com/de0880224936eedca4432ace0e0a872b](https://gyazo.com/de0880224936eedca4432ace0e0a872b)

**Comment by felixoi** [comment]
There should be an UI for looking into the logs. Admins should see the whole log with the IP address, for Devs/Mods the IP address should be hidden.

**Comment by mbax** [comment]
Viewing the logs could look much like what we have now for activity log, but preferably searchable by user, project, IP address, action type (multiple parameters at once?).

Logs of actions on a project should be accessible from that project.
Logs of actions by a user should be accessible from that user.
Logs of actions on a file should be accessible from that file (and project).

Also, because the table schema described was unclear, action logging should be clear in what happened. Like, not just that a user's membership was removed from a group but what group it was and what membership they previously had in that group prior to removal. Basically, before/after states for things. Except project pages/changelogs which I believe we were hoping to move to a separately tracked system with diffs?

**Comment by felixoi** [comment]
I'm working on this @phase. 

@mbax 
Do you want to log if someone uploads a project which upload never finished?
And I think for example when a moderator changes a projects icon we don't need a before and after or?
This diff for pages and changelogs isn't as easy as you think.

**Comment by mbax** [comment]
> Do you want to log if someone uploads a project which upload never finished?

I think the only useful logs here would be errors (which wouldn't be part of this feature) in the process that prevent the user from completing the process.

> And I think for example when a moderator changes a projects icon we don't need a before and after or?

Isn't that a part of Auth, not Ore? If in Ore, could be useful to log who did it.

> This diff for pages and changelogs isn't as easy as you think.

I think it's a difficult challenge, and doesn't need to be part of the first release. A goal for the future! :D

File: conf/evolutions/default/92.sql [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
-  action     TEXT
-);
+ CREATE TABLE user_action_log (
+   id         SERIAL PRIMARY KEY,
+   created_at TIMESTAMP NOT NULL DEFAULT NOW(),
+   user_id    INT REFERENCES users,
+   address    INET NOT NULL,
+   context    VARCHAR(40) NOT NULL,
+   context_id INT,
+   action     VARCHAR(40) NOT NULL,
```
Comment (phase) [inline conf/evolutions/default/92.sql]: Why is this only 40 chars?

Comment (felixoi) [reply conf/evolutions/default/92.sql]: Because it's only a name, it shouldn't be longer than that

**Comment by phase** [comment]
pic of frontend

Backend to frontend communication is done. It doesn't look too pretty. @felixoi I have no idea how to do the pagination so slap that on here.

File: app/controllers/project/Pages.scala [hunk 3]
```diff
@@ -160,15 +160,19 @@ class Pages @Inject()(forms: OreForms,
...
-                  UserActionLogger.log(request.request, UserActionContexts.PROJECT, data.project.id.getOrElse(-1), UserActions.PROJECT_PAGE_EDITED, "", "") //todo data
                   data.project.getOrCreatePage(pageName, parentId, pageData.content)
                 }
               } else {
                 val pageName = pageData.name.getOrElse(parts(0))
                 data.project.getOrCreatePage(pageName, parentId, pageData.content)
               }
               created flatMap { createdPage =>
-                createdPage.setContents(pageData.content.getOrElse(createdPage.contents))
+                if (pageData.content.isDefined) {
```
Comment (Katrix) [inline app/controllers/project/Pages.scala:166]: Any reason for not using fold/getOrElse here?

File: app/controllers/project/Projects.scala [hunk 4]
```diff
@@ -547,7 +548,8 @@ class Projects @Inject()(stats: StatTracker,
         val data = request.data
         val oldName = data.project.name
         this.projects.rename(data.project, newName).map { _ =>
-          UserActionLogger.log(request.request, UserActionContexts.PROJECT, data.project.id.getOrElse(-1), UserActions.PROJECT_RENAME, s"$author/$newName", s"$author/$oldName")
+          service.insert(UserActionLogEntry(None, None, 1, InetString.apply("::1"), GlobalActionContext.Project, 1, UserActions.VersionDeleted, 1, "1", "2"))
```
Comment (Katrix) [inline app/controllers/project/Projects.scala]: Any reason for apply call here?

File: app/models/user/UserActions.scala [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import db.impl.UserActionLogTable
+import db.impl.model.OreModel
+import enumeratum.values.{IntEnum, IntEnumEntry}
+import enumeratum.{EnumEntry, Enum}
+import ore.StatTracker
+import ore.user.UserOwned
+
+case class UserActionLogEntry(override val id: Option[Int] = None,
+                              override val createdAt: Option[Timestamp] = None,
+                              private val _userId: Int,
```
Comment (Katrix) [inline app/models/user/UserActions.scala]: Any reason for hiding the values here if they are then exposed again later? The bytecode is AFAIK identical.

Comment (felixoi) [reply app/models/user/UserActions.scala]: did not touch this

File: app/controllers/project/Projects.scala (hunk 4 above)
Comment (felixoi) [reply app/controllers/project/Projects.scala]: debug statement removed in local changes.

**Pull request merged**